tools/apiview/emitters/cadl-apiview/src/apiview.ts

```diff
@@ -0,0 +1,739 @@
+  }
+
+  private tokenizeDecorators(nodes: readonly DecoratorExpressionNode[], inline: boolean) {
+    const filteredNodes = this.filterDecorators(nodes);
+    if (!inline && filteredNodes.length) {
+      this.blankLines(1);
+    }
+    for (const node of filteredNodes) {
+      NamespaceStack.push(generateId(node)!);
+      this.tokenize(node);
+      if (!inline) {
+        this.lineMarker();
+      }
+      NamespaceStack.pop();
+      if (!inline) {
+        this.newline();
+      }
+    }
+  }
+
+  private getFullyQualifiedIdentifier(node: MemberExpressionNode, suffix?: string): string {
+    switch (node.base.kind) {
+      case SyntaxKind.Identifier:
+        return `${node.base.sv}.${suffix}`;
+      case SyntaxKind.MemberExpression:
+        return this.getFullyQualifiedIdentifier(node.base, `${node.base.id.sv}.${suffix}`);
+    }
+  }
+
+  private tokenizeIdentifer(
+    node: IdentifierNode | MemberExpressionNode | StringLiteralNode,
+    style: "declaration" | "reference" | "member"
+  ) {
+    switch (node.kind) {
+      case SyntaxKind.MemberExpression:
+        const defId = this.getFullyQualifiedIdentifier(node, node.id.sv);
+        switch (style) {
+          case "reference":
+            this.typeReference(defId);
+            break;
+          case "member":
+            this.member(defId);
+            break;
+          case "declaration":
+            throw new Error(`MemberExpression cannot be a "declaration".`);
+        }
+        break;
+      case SyntaxKind.StringLiteral:
+        if (style != "member") {
+          throw new Error(`StringLiteral type can only be a member name. Unexpectedly "${style}"`);
+        }
+        this.member(node.value);
+        break;
+      case SyntaxKind.Identifier:
+        switch (style) {
+          case "declaration":
+            this.typeDeclaration(node.sv, NamespaceStack.value());
+            break;
+          case "reference":
+            const defId = definitionIdFor(node.sv);
+            this.typeReference(node.sv, defId);
+            break;
+          case "member":
+            this.member(node.sv);
+            break;
+        }
+    }
+  }
+
+  private buildNavigation(ns: NamespaceModel) {
+    NamespaceStack.reset();
+    this.navigation(new ApiViewNavigation(ns));
+  }
+
+  resolveMissingTypeReferences() {
+    for (const token of this.Tokens) {
+      if (token.Kind == ApiViewTokenKind.TypeName && token.NavigateToId == "__MISSING__") {
+        token.NavigateToId = definitionIdFor(token.Value!);
+      }
+    }
+  }
+}
+
+export abstract class NamespaceStack {
+  private static stack = Array<string>();
+
+  static push(val: string) {
+    NamespaceStack.stack.push(val);
+  }
+
+  static pop(): string | undefined {
+    return NamespaceStack.stack.pop();
+  }
+
+  static value(): string {
+    return NamespaceStack.stack.join(".");
+  }
+
+  static reset() {
+    this.stack = Array<string>();
+  }
+}
+
+const typeDeclarations = new Set<string>();
```

Comment:
would not put this on the global scope, multiple run of the emitter otherwise could create issues.

Reply:
I originally had it on the ApiViewDocument object, but this results in it being serialized when converted to JSON which makes APIView explode. By multiple runs, do you mean multiple parallel runs? I'm not sure that's a likely scenario, but where would you recommend locating these?

Reply:
What maybe I would do is change ApiViewDocument to be an interface instead of a class. Serializing class in javascript feels like you might end up with unwanted behavior.

Then change the `emit` function to instead be `export function createApiViewDocument(): ApiViewDocument` and have that in there.

Reply:
Thanks @timotheeguerin that worked pretty well I think. Please see the following commit and let me know if you think I could do it differently:
https://github.com/Azure/azure-sdk-tools/pull/4160/commits/6afce8ddce54c422ecaa47df09eff85bdd8c3c56